### DISCUSSION TITLE
STM32H7 and RCC Examples

### DISCUSSION THREAD

**Discussion by thestumbler** [modm-io/modm#1083] [General]
Are there any examples which show RCC configuration of an H7 board using multiple PLLs? I've moved on to making a custom board definition. and am attempting to copy over the initialization settings from a working non-MODM project. If that project was more similar to the Nucleo H723 board example, I would be fine. But it is not. And furthermore, the already complexity of clock control configuration has increased substantially from the F4 to H7 families, and as an H7 beginner it is a little daunting.

Speaking of examples, several times when browsing the example projects, I wished for a different way to view them. I can imagine a way to "sort" the projects by one or two methods besides their associated development boards. One immediately obvious way is by the project name. So you could look for all examples of Ethernet projects, for example. Another sorting would be similar to what's already in the documentation, by project feature -- show me all projects that talk to an accelerometer. That's one kind of feature. Another would be more low-level. In recent days, I was searching haphazardly through the examples for projects using their own custom interrupt handlers, or as noted above, more complicated RCC setups.

It wouldn't be that difficult to manually make a cross reference table for one or two of these ideas. But you guys seem very adept at automatic document generation, perhaps an alternate indexing table like I've described is easy to crank out whenever projects are added to the examples folder.

**Comment by salkinium** [comment]
> Are there any examples which show RCC configuration of an H7 board using multiple PLLs? 

Yes, [here](https://github.com/modm-io/modm/blob/develop/src/modm/board/nucleo_h743zi/board.hpp#L114-L123).
I would recommend to use the CubeMX clock configuration page to figure it out and then copy that into your PLL config.
[I briefly investigated a C++ interface to the clock tree config](https://github.com/salkinium/save-the-clocktrees), but it didn't really work, since it's actually a clock graph, not a tree.

Regarding the example, [we've had a discussion about it too](https://github.com/modm-io/modm/discussions/731), but lacked the motivation to change it…

**Reply by thestumbler** [reply]
By the way, I solved that problem with a double-whammy.

First of all, while I would like to be able to make a MODM project with the same RCC setup as the main application, I realized it is not necessary for the boot loader. I can continue using the RCC setup from the Nucleo examples just fine. 

But there was one other surprise, and it took me a day or two before it hit me. The custom hardware uses a 16 MHz HSE, but the Nucleo uses an 8 MHz one. I thought I was taking that into account (I was), but the custom board would just not run correctly. I was fighting some other unrelated hardware issues at the time which obscured my clock problem, until finally a couple of days ago it hit me. 

Where as my serial port was not sending the expected data, it was sending something, just clearly at a whacko baud rate. I began to question whether my custom hardware really had a 16 MHz crystal installed or not, and I went digging into that. So I really had a 16 MHz crystal, but the Nucleo had a 25 MHz crystal -- not only an unexpected value, but it was NOT POPULATED! That's when it hit me what was going on.

The Nucleo board uses the attached STLink module's crystal-controlled oscillator output as an external clock signal over the HSE input pin. The setup in the example code enables HSE, but "enable" doesn't mean enable the oscillator circuit. That's assigned to the HSE_BYPASS bit, which is really the HSE_OSCILLATOR_ENABLE bit in disguise. I did a bit more head scratching when I couldn't find the text "BYPASS" anywhere in the entire MODM repository, then I saw that ST had abbreviated it something like BYP. That was all I needed to learn that I should call `enableExternalCrystal()` instead of `enableExternalClock()` and it should work as expected.

**Reply by salkinium** [reply]
Yeah… we really need an open-source clock configuration UI for microcontrollers. The H7 is a real challenge with its multiple power and clock domains.
